tracer/src/Datadog.Trace/Activity/Handlers/DefaultActivityHandler.cs

```diff
@@ -102,18 +108,26 @@ public void ActivityStarted<T>(string sourceName, T activity)
                     return;
                 }
 
-                ActivityMappingById.GetOrAdd(activity.Id, _ => new(activity.Instance, CreateScopeFromActivity(activity, parent, traceId, spanId, rawTraceId, rawSpanId)));
+                Log.Information("GetOrAdd for {Name} with ID: {ID}", activity.OperationName, activity.Id);
+
+                if (isRemoteContext)
+                {
+                    ActivityMappingById.GetOrAdd(activity.ParentId!, _ => new(activity.Instance, CreateScopeFromActivity(activity, parent, traceId, spanId, rawTraceId, rawSpanId)));
```

Comment:
I don't follow why this is correct 🤔 don't we still need to track the ID of the current activity?

Reply:
Hmm I think you might be right, I'll test out what happens for `ActivityContext` -> `ParentActivity` -> `ChildActivity`.

But for why I'm mapping to `activity.ParentId` it seems that when an `ActivityContext` is supplied the `ParentId` for all of the `Activity` objects with that context is the combination of `ActivityContext.TraceId` + `ActivityContext.SpanId`

Whereas their own `Activity.Id` is the combination of `ActivityContext.TraceId` + `Activity.SpanId` so all of the mappings for them end up failing without mapping by `ParentId`

Reply:
I'm still not quite following 😄 

> it seems that when an `ActivityContext` is supplied, the `ParentId` for all of the `Activity` objects with that context is the combination of `ActivityContext.TraceId` + `ActivityContext.SpanId`

That's what I'd expect I think. As I understand it from this code:

`Activity.Id` = `Activity.TraceId` + `Activity.SpanId`

The `ActivityContext` represents a "fictious/remote" activity, so the `ID` of the _context_ should be `ActivityContext.TraceId` + `ActivityContext.SpanId`, and hence this is also the parent ID

Each of the spans created from this `ActivityContext` _should_ have the same `ParentId`. But they will have different `Id`s. And it's the _Id_ that we need to look up by I think?

Otherwise this (based on your PR example):

```csharp
ActivitySource source  = new ActivitySource("Name");
var someContext = new ActivityContext(ActivityTraceId.CreateRandom(), ActivitySpanId.CreateRandom(), ActivityTraceFlags.None);

using var firstActivity = source.StartActivity("First", ActivityKind.Server, someContext);
using var secondActivity = source.StartActivity("Second", ActivityKind.Server, someContext);
```

Will be problematic. When the `secondActivity` is disposed, we'll look up by the `ParentId`, and call `TryRemove()`. When we then do the same for `firstActivity`, we won't have a reference any more, so can't call `CloseActivityScope`? So I think we would "leak" spans?

Reply:
Hmm, I'll have to take another look at this as maybe I'm just missing something or making this too difficult than what it is.

But, for the `secondActivity` it's `Activity.Id` is in the mappings so we should find that first. The `firstActivity` is in the mappings, but is mapped via its `Activity.ParentId` instead of its `Activity.Id`. So I don't think we'd leak anything?

Reply:
We went through this offline - changed to not need to use the `ParentId`